# sanger-tol/treeval: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels, P. et al. 2020. ‘The NF-core framework for community-curated bioinformatics pipelines’, Nature Biotechnology, 38(3), pp. 276–278. doi:10.1038/s41587-020-0439-x.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso, P. et al. 2017. ‘Nextflow enables reproducible computational workflows’, Nature Biotechnology, 35(4), pp. 316–319. doi:10.1038/nbt.3820.

## Pipeline tools

- [Bedtools](https://bedtools.readthedocs.io/en/latest/)

  > Quinlan, A.R. and Hall, I.M. 2010. ‘BEDTools: A flexible suite of utilities for comparing genomic features’, Bioinformatics, 26(6), pp. 841–842. doi:10.1093/bioinformatics/btq033.

- [BUSCO](https://busco.ezlab.org)

  > Manni, M. et al. 2021. Busco update: Novel and streamlined workflows along with broader and deeper phylogenetic coverage for scoring of eukaryotic, prokaryotic, and viral genomes, Molecular biology and evolution. Available at: https://pubmed.ncbi.nlm.nih.gov/34320186/ (Accessed: 22 June 2023).

- [bwa-mem2](https://ieeexplore.ieee.org/document/8820962)

  > Vasimuddin, Md. et al. 2019. ‘Efficient architecture-aware acceleration of BWA-mem for multicore systems’, 2019 IEEE International Parallel and Distributed Processing Symposium (IPDPS) [Preprint]. doi:10.1109/ipdps.2019.00041.

- [Cooler](https://github.com/open2c/cooler)

  > Abdennur, N. and Mirny, L.A. 2019. ‘Cooler: Scalable storage for hi-C data and other genomically labeled arrays’, Bioinformatics, 36(1), pp. 311–316. doi:10.1093/bioinformatics/btz540.

- [Find Telomere](https://github.com/VGP/vgp-assembly/tree/master/pipeline/telomere)

  > VGP. 2022. vgp-assembly telomere [online]. https://github.com/VGP/vgp-assembly/tree/master/pipeline/telomere. (Accessed on 28th February 2023).

- [Juicer](https://github.com/aidenlab/juicer)

  > Durand, N.C. et al. 2016. ‘Juicer provides a one-click system for analyzing loop-resolution hi-C experiments’, Cell Systems, 3(1), pp. 95–98. doi:10.1016/j.cels.2016.07.002.

- [Minimap2](https://pubmed.ncbi.nlm.nih.gov/34623391/)

  > Li, H. 2021. ‘New strategies to improve MINIMAP2 alignment accuracy’, Bioinformatics, 37(23), pp. 4572–4574. doi:10.1093/bioinformatics/btab705.

- [Miniprot](https://arxiv.org/abs/2210.08052)

  > Li, H. 2022. Protein-to-genome alignment with miniprot (v2). Arxiv [e-journal]. (Accessed on 25th January 2023). doi: 10.48550/arXiv.2210.08052.

- [Mummer](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005944)

  > Marçais, G. et al. 2018. ‘Mummer4: A fast and versatile genome alignment system’, PLOS Computational Biology, 14(1). doi:10.1371/journal.pcbi.1005944.

- [Pandas](https://pandas.pydata.org/)

  > Redback, J. 2022. pandas-dev/pandas: Pandas 1.4.3 [online]. Zenodo. doi: 10.5281/zenodo.6702671. (Accessed on 28th February 2023).

- [Perl](https://perldoc.perl.org/perl)

  > Perl Organisation. 2023. Perl Language Reference v5.36.0. https://perldoc.perl.org/perl. (Accessed 28th February 2023).

- [PretextMap](https://github.com/wtsi-hpag/PretextMap)

  > Harry, E. 2022. PretextView [online]. https://github.com/wtsi-hpag/PretextView. (Accessed on 7th June 2023).

- [Pybedtools](https://github.com/daler/pybedtools)

  > Daler. 2022. pybedtools [online]. https://github.com/daler/pybedtools. (Accessed on 7th June 2023).

- [Python: 3.10](https://docs.python.org/3.10/whatsnew/3.10.html)

  > Python Software Foundation. 2023. Python Language Reference v3.10. https://docs.python.org/3.10/whatsnew/3.10.html. (Accessed 28th February 2023).

- [PyFasta](https://github.com/brentp/pyfasta/)

  > Brentp. 2018. pyfasta [online]. http://github.com/brentp/pyfasta/. (Accessed on 7th June 2023).

- [Samtools](https://pubmed.ncbi.nlm.nih.gov/33590861/)

  > Di Tommaso, Paolo, et al. 2017. “Nextflow Enables Reproducible Computational Workflows.” Nature Biotechnology, 35(4), pp. 316–19, https://doi.org/10.1038/nbt.3820.

- [SeqTK](https://github.com/lh3/seqtk)

  > Li, Heng. 2023. seqtk [online]. https://github.com/lh3/seqtk. (Accessed on 7th June 2023).

- [staden_io_lib / iolib](https://github.com/jkbonfield/io_lib)

  > Bonfield JK. 2023. io_lib [online]. https://github.com/jkbonfield/io_lib. (Accessed on 7th June 2023).

- [Tabix](http://www.htslib.org/doc/tabix.html)

  > Li, Heng. 2023. tabix [online]. http://www.htslib.org/doc/tabix.html. (Accessed on 7th June 2023).

- [UCSC tools](https://github.com/ucscGenomeBrowser/kent/tree/master)

  > UCSC Genome Browser Group. 2023. kent [online]. https://github.com/ucscGenomeBrowser/kent/tree/master. (Accessed on 7th June 2023).

- [WindowMasker](https://pubmed.ncbi.nlm.nih.gov/16287941/)

  > Morgulis, A., et al. 2006. WindowMasker: window-based masker for sequenced genomes. Bioinformatics. 22(2). pp.134–141. doi: 10.1093/bioinformatics/bti774.

- [lep_busco_painter](https://www.biorxiv.org/content/10.1101/2023.05.12.540473v1.full.pdf)

  > Wright, C. et al. 2023. Chromosome evolution in Lepidoptera. bioRxiv. 540473. https://doi.org/10.1101/2023.05.12.540473

- [Java](https://docs.oracle.com/javase/8/docs/api/overview-summary.html)

  > Oracle. 2023. Java Documentation. https://docs.oracle.com/javase/8/docs/index.html. (Accessed on 25th September 2023).

- [coreutils](https://github.com/coreutils/coreutils)

  > GNU Coreutils. 2023. coreutils [online]. https://github.com/coreutils/coreutils/releases/tag/v9.4. (Accessed on 25th September 2023).

## Software packaging/containerisation tools

- [Conda](https://conda.org/)

  > conda contributors. conda: A system-level, binary package and environment manager running on all major operating systems and platforms. Computer software. https://github.com/conda/conda

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Di Tommaso, Paolo, et al. 2017. “Nextflow Enables Reproducible Computational Workflows.” Nature Biotechnology, 35(4), pp. 316–19, https://doi.org/10.1038/nbt.3820.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > Grüning, Björn, et al. 2018. “Bioconda: sustainable and comprehensive software distribution for the life sciences". Nature Methods, 15, pp. 475-6, https://doi.org/10.1038/s41592-018-0046-7.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, Dirk, et al. 2014. “Docker: Lightweight Linux Containers for Consistent Development and Deployment.". Association for Computing Machinery. 2014(239)

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)
  > Kurtzer, Gregory M., et al. 2017. “Singularity: Scientific containers for mobility of compute.", PLOS ONE, 12(5), pp. e0177459, https://doi.org/10.1371/journal.pone.0177459.
